File: product_of_self.java
// Time Complexity :O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english

// Your code here along with comments explaining your approach

public int[] productExceptSelf(int[] nums) {
    if(nums == null || nums.length == 0) return new int[0];
    int[] ans = new int[nums.length]; //output array
    int left = 1;
    //for every index, find the product of elements to it's left and to it's right and multiply them
    //to get the output result

    //Calculating the product of the left side elements for every element at that index. 
    for (int i = 0; i < nums.length; i++) {
        ans[i] = left;
        left *= nums[i];
    }
    
    //we can avoid taking an extra right array as the elements in the right array are results of
    //an individual operstion. We don't need previous values to calculate the current element of the right
    //array
    int right = 1;
    for (int i = nums.length - 1; i >= 0; i--) {
        ans[i] *= right;
        right *= nums[i];
    }
    
    return ans;   
}

File: product of self.py
# // Time Complexity : O(n)
# // Space Complexity : O(n) but O(1) in leetcode, since ans array does not take into account as space complexity
# // Did this code successfully run on Leetcode :
# // Three line explanation of solution in plain english:
# create an ans array of nums lenght and assign first element as 1 and multiply
# i-1 elements of ans and nums until end - 1. Assign end as 1 and similary multiply
# i element of nums from end to start.   

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        length = len(nums)
        
        r = ans = [0] * length
        
        # left end element is 1, since left to first element is none
        ans[0] = 1
        
        # product of ans and nums of previous element and store it in current index of ans
        for i in range(1,length):
            ans[i] = ans[i-1]*nums[i-1]
        
        # right end element is 1, since right to first element is none
        r= 1
        
        # product of ans and nums of next element and store in place in ans array
        for i in reversed(range(length)):
            ans[i] = ans[i] * r
            
            # update r from right end
            r *= nums[i]
            
        return ans